pkg/bpf/collection.go

```diff
@@ -41,6 +45,110 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
 	return spec, nil
 }
 
+func removeUnreachableTailcalls(spec *ebpf.CollectionSpec) error {
+	type TailCall struct {
+		referenced bool
+		visited    bool
+		spec       *ebpf.ProgramSpec
+	}
+
+	entrypoints := make([]*ebpf.ProgramSpec, 0)
+	tailcalls := make(map[uint32]*TailCall)
+
+	for _, prog := range spec.Programs {
+		if prog.SectionName == "tc" || prog.SectionName == "xdp" {
+			entrypoints = append(entrypoints, prog)
+			continue
+		}
+
+		var id, slot uint32
+		if _, err := fmt.Sscanf(prog.SectionName, "%d/%v", &id, &slot); err != nil {
+			continue
+		}
+
+		const (
+			cilium_policy_map        = 1
+			cilium_calls_map         = 2
+			cilium_egress_policy_map = 4
+		)
+
+		// Any programs in the cilium_policy_map or cilium_egress_policy_map are
+		// called dynamically, assume they are reachable and treat as entrypoints.
+		if id == cilium_policy_map || id == cilium_egress_policy_map {
+			entrypoints = append(entrypoints, prog)
+			continue
+		}
+
+		// Any programs not in the tail call map can be ignored (always load them).
+		if id != cilium_calls_map {
+			continue
+		}
+
+		tailcalls[slot] = &TailCall{
+			spec: prog,
+		}
+	}
+
+	// Discover all tailcalls that are reachable from the given program.
+	visit := func(prog *ebpf.ProgramSpec, tailcalls map[uint32]*TailCall) {
+		for i := 0; i < len(prog.Instructions); i++ {
+			inst := prog.Instructions[i]
+			if !(inst.IsBuiltinCall() && inst.Constant == int64(asm.FnTailCall)) {
+				continue
+			}
+
+			if i-1 <= 0 {
+				continue
+			}
+
+			// For all tail calls where the index is static we use the `tail_call_static` function.
+			// This function uses inline assembly to guarantee that the instruction right above the
+			// tail call instruction is a `mov r3, <index>` instruction. The IMM value of this
+			// instruction is used as the index for the tail call.
+			prev := prog.Instructions[i-1]
+			if !(prev.OpCode.Class() == asm.ALU64Class && prev.Dst == asm.R3) {
+				continue
+			}
+
+			tc := tailcalls[uint32(prev.Constant)]
+			if tc == nil {
+				// This tends to happen due to the custom tail calls map which also uses
```

Comment:
I think you can retrieve the name of the map from the `ldwimm64` used to load the map fd, no? Then we could actually check here that this references the correct map. As is is right now I'm a bit worried that this is too loose.

Reply:
:thinking: , yea, it should have map reference. I will try this, see if I can make it work.

Reply:
So yes and no. It isn't as easy as just looking at 2 instructions above the tail call. That instruction is always `Mov R2, Rx`. And then its up to the compiler to pick a register and make sure it contains the fd. So far it always appears to go:
```
Mov Rx, <map fd>
Mov R1, Rctx
Mov R2, Rx
Mov R3, <index>
Call tail_call
```

So I wrote some code that looks for this. However, if the compiler ever decides to use a stack slot for this instead of re-getting the map fd we are strewed. While I think this is unlikely, if I am wrong, we will get random missing tail-calls.

For now I want to see if this works in CI. If not then we have to revert to not knowing, or perhaps change `tail_call_static` to emit `Mov R2, <map fd>` instead (if that is possible). 

Last option is that if we are unable to find the `Mov Rx, <map fd>`, that we assume the worst case and mark the index as referenced, that way we load a dead program instead of having missing tail calls.

Reply:
Weird, why does the compiler emit the extra `mov Rx`?

Anyways, can we somehow make the compiler emit a relocation for the `mov r2`? Maybe we can use the `CONFIG` macro hack?

```
asm volatile("%[out] = __config_" #name " ll" : [out]"=r"(out));
```

(Might also be able to use the ll suffix to skip the separate load?)

Reply:
I managed to get this to work :partying_face: